source/extensions/filters/network/rbac/rbac_filter.cc

```diff
@@ -69,11 +71,47 @@ RoleBasedAccessControlFilterConfig::RoleBasedAccessControlFilterConfig(
       enforcement_type_(proto_config.enforcement_type()),
       delay_deny_ms_(PROTOBUF_GET_MS_OR_DEFAULT(proto_config, delay_deny, 0)) {}
 
+void RoleBasedAccessControlFilter::initializeReadFilterCallbacks(
+    Network::ReadFilterCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+  callbacks_->connection().addConnectionCallbacks(*this);
+
+  callbacks_->connection().streamInfo().filterState()->setData(
+      Envoy::TcpProxy::ReceiveBeforeConnectKey,
+      std::make_unique<StreamInfo::BoolAccessorImpl>(true),
+      StreamInfo::FilterState::StateType::ReadOnly, StreamInfo::FilterState::LifeSpan::Connection);
+}
+
+Network::FilterStatus RoleBasedAccessControlFilter::onNewConnection() {
+  // We stop iteration so that the tcp_proxy doesn't make a connection upstream until we have a
+  // chance to make an RBAC decision in onEvent or onData.
+  return Network::FilterStatus::StopIteration;
+}
+
 Network::FilterStatus RoleBasedAccessControlFilter::onData(Buffer::Instance&, bool) {
   if (is_delay_denied_) {
     return Network::FilterStatus::StopIteration;
   }
 
+  if (runAuthorization()) {
+    return Network::FilterStatus::Continue;
+  } else {
+    return Network::FilterStatus::StopIteration;
+  }
+}
+
+void RoleBasedAccessControlFilter::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    resetTimerState();
+  } else if (event == Network::ConnectionEvent::Connected) {
+    if (runAuthorization()) {
+      callbacks_->continueReading();
```

Comment:
This would be good to guard with onNewConnection - I'm not sure the ordering is guaranteed.